sns/src/main/scala/akka/stream/alpakka/sns/SnsPublishFlowStage.scala

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.alpakka.sns
+
+import akka.stream._
+import akka.stream.stage._
+import com.amazonaws.handlers.AsyncHandler
+import com.amazonaws.services.sns.AmazonSNSAsync
+import com.amazonaws.services.sns.model.{PublishRequest, PublishResult}
+
+final class SnsPublishFlowStage(topicArn: String, snsClient: AmazonSNSAsync)
+    extends GraphStage[FlowShape[String, PublishResult]] {
+
+  private val in = Inlet[String]("SnsPublishFlow.in")
+  private val out = Outlet[PublishResult]("SnsPublishFlow.out")
+
+  override def shape: FlowShape[String, PublishResult] = FlowShape.of(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with InHandler with OutHandler with StageLogging {
+
+      private var inFlight = 0
+      private val failureCallback = getAsyncCallback[Throwable](handleFailure)
+      private val successCallback = getAsyncCallback[PublishResult](handleSuccess)
+
+      private def handleFailure(ex: Throwable): Unit =
+        failStage(ex)
+
+      private def handleSuccess(result: PublishResult): Unit = {
+        log.debug("Published SNS message: {}", result.getMessageId)
+        inFlight -= 1
+        if (isAvailable(out)) {
+          if (!hasBeenPulled(in)) tryPull(in)
```

Comment:
Wouldn't this limit the Flow to process elements only one by one? Otherwise we would always have to wait for the output port to pull?

Thats why I also made inFlight as an Int and not a Boolean.